+precice/@Participant/private/preciceGateway.cpp

```diff
@@ -0,0 +1,353 @@
+                break;
+            }
+            case FunctionID::writeData:
+            {
+                const std::string meshName = convertToString(inputs[1]);
+                const std::string dataName = convertToString(inputs[2]);
+                const std::vector<int> vertexIDs = convertToIntVector(inputs[3]);
+                const std::vector<double> values = convertToDoubleVector(inputs[4]);
+                interface->writeData(meshName,dataName,vertexIDs,values);
+                break;
+            }
+            case FunctionID::readData:
+            {
+                const std::string meshName = convertToString(inputs[1]);
+                const std::string dataName = convertToString(inputs[2]);
+                const TypedArray<int32_t> size = inputs[3];
+                const std::vector<int32_t> vertexIDs = convertToIntVector(inputs[4]);
+                size_t dim = interface->getDataDimensions(meshName,dataName);
+                const TypedArray<double> relativeReadTime = inputs[5];
+                std::vector<double> values(size[0]*dim);
+                interface->readData(meshName,dataName,vertexIDs,relativeReadTime[0],values);
+                outputs[0] = factory.createArray<double>({1, size[0]*dim}, values.data(), values.data()+values.size());
+                break;
+            }
+            case FunctionID::requiresGradientDataFor:
+            {
+                const std::string meshName = convertToString(inputs[1]);
+                const std::string dataName = convertToString(inputs[2]);
+                bool requiresGradientData = interface->requiresGradientDataFor(meshName,dataName);
+                outputs[0] = factory.createScalar<bool>(requiresGradientData);
+                break;
+            }
+            case FunctionID::writeGradientData:
+            {
+                const std::string meshName = convertToString(inputs[1]);
+                const std::string dataName = convertToString(inputs[2]);
+                const std::vector<int32_t> vertexIDs = convertToIntVector(inputs[4]);
+                const std::vector<double> gradientValues = convertToDoubleVector(inputs[5]);
+                interface->writeGradientData(meshName,dataName, vertexIDs, gradientValues);
+                break;
+            }
+            case FunctionID::setMeshAccessRegion:
+            {
+                const std::string meshName = convertToString(inputs[1]);
+                const std::vector<double> boundingBox = convertToDoubleVector(inputs[2]);
+                interface->setMeshAccessRegion(meshName,boundingBox);
+                break;
+            }
+            case FunctionID::getMeshVertexIDsAndCoordinates:
+            {
+                const std::string meshName = convertToString(inputs[1]);
+                const TypedArray<int32_t> size = inputs[2];
+                std::vector <int32_t> ids(size[0]);
+                std::vector <double> positions(size[0]*3);
```

Comment:
Where does this 3 come from? I think this should be the return value of `getMeshDimensions()`?!